src/unix/exec.c

```diff
@@ -36,8 +36,8 @@ static void build_exec_stack(process p, thread t, Elf64_Ehdr * e, void *start,
 {
     exec_debug("build_exec_stack start %p, tid %d, va 0x%lx\n", start, t->tid, va);
 
-    /* allocate process stack at top of first 2gb of address space */
-    u64 stack_start = 0x80000000 - PROCESS_STACK_SIZE;
+    /* allocate process stack at top of 32-bit address space */
+    u64 stack_start = 0x100000000 - PROCESS_STACK_SIZE;
```

Comment:
The [0x80000000, 0x100000000) address range is used for mmap() with the MAP_32BIT flag, so if part of this range is used for program segments or the stack, this needs to be reflected in the p->virtual32 heap (on x86).

Reply:
Thanks for pointing that out! I added the elf load segments to the p->virtual32 heap for x86.